src/sentry/auth/access.py

```diff
@@ -187,23 +188,31 @@ def to_django_context(self) -> Mapping[str, bool]:
 
 
 class OrganizationGlobalAccess(Access):
-    def __init__(self, organization: Organization):
+    def __init__(self, organization: Organization, scopes: Iterable[str], **kwargs):
+        # TODO: Refactor these into cached properties to avoid needless queries
+        teams = frozenset(Team.objects.filter(organization=organization, status=TeamStatus.VISIBLE))
+        projects = frozenset(
+            Project.objects.filter(organization=organization, status=ProjectStatus.VISIBLE)
+        )
+
         super().__init__(
-            sso_is_valid=True,
             is_active=True,
-            has_global_access=True,
-            scopes=frozenset(settings.SENTRY_SCOPES),
+            scopes=frozenset(scopes),
             organization_id=organization.id,
+            teams=teams,
+            projects=projects,
+            **kwargs,
         )
 
     def has_team_access(self, team: Team) -> bool:
-        return team.organization_id == self.organization_id
+        return team.organization_id == self.organization_id and (
+            self.has_global_access or team.status == TeamStatus.VISIBLE
+        )
 
     def has_project_access(self, project: Project) -> bool:
-        return project.organization_id == self.organization_id
-
-    def has_scope(self, scope: str) -> bool:
-        return True
+        return project.organization_id == self.organization_id and (
+            self.has_global_access or project.status == ProjectStatus.VISIBLE
+        )
```

Comment:
Calling the class "Organization *Global* Access" when `has_global_access` is sometimes false is 🤢. But I think the salient thing we're representing is that it represents access to the whole organization and not a particular subset of teams and projects. It's a minor quibble that "true" global access means you have access to deleted objects.

I really didn't want to change this class's `has_*_access` bodies (the change came late in refactoring) but, as mentioned in the PR description, the unit tests were insistent. Ultimately, I think it's fine; this is just carefully replicating the behavior of the base method without needing to look at a set:
```python
        if not self.is_active:
            return False
        if self.has_global_access and self.organization_id == project.organization_id:
            return True
        return project in self.projects
```

When that set is eventually a cached property, it would pay off with a performance gain of never needing to initialize the set at all.

Reply:
Wondering whether `has_global_access` should renamed here to something like `has_deletion_access` to make this clearer? Or maybe in the future this difference can just be moved over to a scope